File: algorithms/algorithms-cplusplus/leetcode/Question_0897_Increasing_Order_Search_Tree.cc

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  TreeNode* increasingBST(TreeNode* root) {
    if (!root) {
      return root;
    }
    TreeNode res(0);
    TreeNode* p = &res;
    inorder(root, p);
    return res.right;
  }

 private:
  void inorder(TreeNode* root, TreeNode*& cur) {
    if (!root) {
      return;
    }
    inorder(root->left, cur);
    root->left = nullptr;
    cur->right = root;
    cur = root;
    inorder(root->right, cur);
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0897_Increasing_Order_Search_Tree.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0897_Increasing_Order_Search_Tree {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private TreeNode cur;

    public TreeNode increasingBST(TreeNode root) {
        TreeNode res = new TreeNode(0);
        cur = res;
        inorder(root);
        return res.right;
    }

    public void inorder(TreeNode node) {
        if (node == null) {
            return;
        }
        inorder(node.left);
        node.left = null;
        cur.right = node;
        cur = node;
        inorder(node.right);
    }
}
